runner/logs.go

```diff
@@ -82,6 +133,157 @@ func downloadLogsFromOneServer(wg *sync.WaitGroup, server *state.Server, globalC
 	}
 }
 
+func downloadLogsForServer(server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) (state.PersistedLogState, bool, error) {
+	grant, err := getLogsGrant(server, globalCollectionOpts, logger)
+	if err != nil || !grant.Valid {
+		return server.LogPrevState, false, err
+	}
+
+	transientLogState := state.TransientLogState{CollectedAt: time.Now()}
+	defer transientLogState.Cleanup()
+
+	var newLogState state.PersistedLogState
+	newLogState, transientLogState.LogFiles, transientLogState.QuerySamples, err = system.DownloadLogFiles(server.LogPrevState, server.Config, logger)
+	if err != nil {
+		return newLogState, false, errors.Wrap(err, "could not collect logs")
+	}
+
+	err = postprocessAndSendLogs(server, globalCollectionOpts, logger, transientLogState, grant)
+	if err != nil {
+		return newLogState, false, err
+	}
+	return newLogState, true, nil
+}
+
+func setupLogStreamer(ctx context.Context, wg *sync.WaitGroup, globalCollectionOpts state.CollectionOpts, logger *util.Logger, servers []*state.Server, logTestSucceeded chan<- bool, logTestFunc func(s *state.Server, lf state.LogFile, lt chan<- bool)) chan state.ParsedLogStreamItem {
+	parsedLogStream := make(chan state.ParsedLogStreamItem, state.LogStreamBufferLen)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		logLinesByServer := make(map[config.ServerIdentifier][]state.LogLine)
+
+		ticker := time.NewTicker(LogStreamingInterval)
+
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case t := <-ticker.C:
+				for identifier := range logLinesByServer {
+					if len(logLinesByServer[identifier]) > 0 {
+						server := &state.Server{}
+						for _, s := range servers {
+							if s.Config.Identifier == identifier {
+								server = s
+							}
+						}
```

Comment:
NBD but since this is just mechanical, we could extract this into a helper to take the clutter out of the function body here:

```golang
func findServerById(servers []*state.Server, id string) {
  for _, s := range servers {
    if s.Config.Identifier == identifier {
      return s
    }
  }
}
```